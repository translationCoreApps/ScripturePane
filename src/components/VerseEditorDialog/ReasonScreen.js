import React from 'react';
import PropTypes from 'prop-types';
import ReasonCheckbox from './ReasonCheckbox';

const styles = {
  root: {
    display: 'flex'
  },
  content: {
    display: 'flex',
    flexGrow: 1
  },
  column: {
    flexGrow: 1
  },
  spacer: {
    flexGrow: 2
  }
};

/**
 * Renders checkboxes for the user to indicate the reason for the change
 */
class ReasonScreen extends React.Component {
  constructor(props) {
    super(props);
    this._handleCheck = this._handleCheck.bind(this);
    this.state = {
      selectedReasons: {}
    };
  }

  /**
   * Checks if a checkbox is selected
   * @param {string} reason
   * @param {bool} checked
   * @return {boolean}
   * @private
   */
  _handleCheck(reason, checked) {
    const {selectedReasons, onChange} = this.props;
    let newReasons = [
      ...selectedReasons
    ];
    if(checked && !newReasons.includes(reason)) {
      newReasons.push(reason);
    } else if(!checked && newReasons.includes(reason)) {
      const index = newReasons.indexOf(reason);
      newReasons.splice(index, 1);
    }
    onChange(newReasons);
  }

  render() {
    const {translate, selectedReasons} = this.props;

    return (
      <div style={styles.root}>
        <div style={styles.spacer}/>
        <div style={styles.content}>
          <div style={styles.column}>
            <ReasonCheckbox reason="spelling"
                            label={translate('editor.spelling')}
                            onCheck={this._handleCheck}
                            selectedReasons={selectedReasons}/>
            <ReasonCheckbox reason="punctuation"
                            label={translate('editor.punctuation')}
                            onCheck={this._handleCheck}
                            selectedReasons={selectedReasons}/>
            <ReasonCheckbox reason="word_choice"
                            label={translate('editor.word_choice')}
                            onCheck={this._handleCheck}
                            selectedReasons={selectedReasons}/>
          </div>
          <div style={styles.column}>
            <ReasonCheckbox reason="meaning"
                            label={translate('editor.meaning')}
                            onCheck={this._handleCheck}
                            selectedReasons={selectedReasons}/>
            <ReasonCheckbox reason="grammar"
                            label={translate('editor.grammar')}
                            onCheck={this._handleCheck}
                            selectedReasons={selectedReasons}/>
            <ReasonCheckbox reason="other"
                            label={translate('editor.other')}
                            onCheck={this._handleCheck}
                            selectedReasons={selectedReasons}/>
          </div>
        </div>
        <div style={styles.spacer}/>
      </div>

    );
  }
}

ReasonScreen.propTypes = {
  onChange: PropTypes.func.isRequired,
  translate: PropTypes.func.isRequired,
  selectedReasons: PropTypes.arrayOf(PropTypes.string).isRequired
};

export default ReasonScreen;
