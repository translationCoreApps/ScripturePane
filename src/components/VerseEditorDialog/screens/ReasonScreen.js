import React from 'react';
import PropTypes from 'prop-types';
import Checkbox from 'material-ui/Checkbox';

const styles = {
  root: {
    display: 'flex'
  },
  content: {
    display: 'flex',
    flexGrow: 1
  },
  column: {
    flexGrow: 1
  },
  checkbox: {
    marginBottom: 16
  },
  spacer: {
    flexGrow: 2
  }
};

/**
 * Renders a reason checkbox
 * @param {string} reason - the reason for the edit
 * @param {string} label - the checkbox label
 * @param {string[]} selectedReasons - an array of selected reasons
 * @param {func} onCheck - callback when the checkbox is changed
 * @return {*}
 * @constructor
 */
class ReasonCheckbox extends React.Component {

  constructor(props) {
     super(props);
     this._handleCheck = this._handleCheck.bind(this);
  }

  _handleCheck(e, checked) {
    const {reason, onCheck} = this.props;
    onCheck(reason, checked);
  }

  render() {
    const {reason, label, selectedReasons} =this.props;
    return (
      <Checkbox style={styles.checkbox}
                checked={selectedReasons.includes(reason)}
                onCheck={this._handleCheck}
                label={label}/>
    );
  }
}

ReasonCheckbox.propTypes = {
  reason: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  selectedReasons: PropTypes.arrayOf(PropTypes.string).isRequired,
  onCheck: PropTypes.func.isRequired
};

/**
 * Renders checkboxes for the user to indicate the reason for the change
 */
class ReasonScreen extends React.Component {
  constructor(props) {
    super(props);
    this._handleCheck = this._handleCheck.bind(this);
    this.state = {
      selectedReasons: {}
    };
  }

  /**
   * Checks if a checkbox is selected
   * @param {string} reason
   * @param {bool} checked
   * @return {boolean}
   * @private
   */
  _handleCheck(reason, checked) {
    const {selectedReasons, onChange} = this.props;
    let newReasons = [
      ...selectedReasons
    ];
    if(checked && !newReasons.includes(reason)) {
      newReasons.push(reason);
    } else if(!checked && newReasons.includes(reason)) {
      const index = newReasons.indexOf(reason);
      newReasons.splice(index, 1);
    }
    onChange(newReasons);
  }

  render() {
    const {translate, selectedReasons} = this.props;

    return (
      <div style={styles.root}>
        <div style={styles.spacer}/>
        <div style={styles.content}>
          <div style={styles.column}>
            <ReasonCheckbox reason="spelling"
                            label={translate('editor.spelling')}
                            onCheck={this._handleCheck}
                            selectedReasons={selectedReasons}/>
            <ReasonCheckbox reason="punctuation"
                            label={translate('editor.punctuation')}
                            onCheck={this._handleCheck}
                            selectedReasons={selectedReasons}/>
            <ReasonCheckbox reason="word_choice"
                            label={translate('editor.word_choice')}
                            onCheck={this._handleCheck}
                            selectedReasons={selectedReasons}/>
          </div>
          <div style={styles.column}>
            <ReasonCheckbox reason="meaning"
                            label={translate('editor.meaning')}
                            onCheck={this._handleCheck}
                            selectedReasons={selectedReasons}/>
            <ReasonCheckbox reason="grammar"
                            label={translate('editor.grammar')}
                            onCheck={this._handleCheck}
                            selectedReasons={selectedReasons}/>
            <ReasonCheckbox reason="other"
                            label={translate('editor.other')}
                            onCheck={this._handleCheck}
                            selectedReasons={selectedReasons}/>
          </div>
        </div>
        <div style={styles.spacer}/>
      </div>

    );
  }
}

ReasonScreen.propTypes = {
  onChange: PropTypes.func.isRequired,
  translate: PropTypes.func.isRequired,
  selectedReasons: PropTypes.arrayOf(PropTypes.string).isRequired
};

export default ReasonScreen;
