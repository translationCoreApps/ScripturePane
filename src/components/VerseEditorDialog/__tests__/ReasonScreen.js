/* eslint-env jest */

jest.mock('../ReasonCheckbox');
import React from 'react';
import ReasonScreen, {updateReasons} from '../ReasonScreen';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import {render} from 'enzyme';

describe('<ReasonScreen/>', () => {
  it('has no selections', () => {
    const wrapper = render(
      <MuiThemeProvider>
        <ReasonScreen selectedReasons={[]} onChange={jest.fn()} translate={k=>k}/>
      </MuiThemeProvider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('has selections', () => {
    const wrapper = render(
      <MuiThemeProvider>
        <ReasonScreen selectedReasons={['spelling']} onChange={jest.fn()} translate={k=>k}/>
      </MuiThemeProvider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});

describe('updateReasons',  () => {
  it('is newly checked', () => {
    const reasons = updateReasons('spelling', true, ['context']);
    expect(reasons).toEqual(['context', 'spelling']);
  });
  it('is already checked', () => {
    const reasons = updateReasons('spelling', true, ['context', 'spelling']);
    expect(reasons).toEqual(['context', 'spelling']);
  });
  it('is newly un-checked', () => {
    const reasons = updateReasons('spelling', false, ['spelling', 'context']);
    expect(reasons).toEqual(['context']);
  });
  it('is already un-checked', () => {
    const reasons = updateReasons('spelling', false, ['context']);
    expect(reasons).toEqual(['context']);
  });
  it('is the only checked', () => {
    const reasons = updateReasons('spelling', true, []);
    expect(reasons).toEqual(['spelling']);
  });
  it('is completely un-checked', () => {
    const reasons = updateReasons('spelling', false, []);
    expect(reasons).toEqual([]);
  });
});
