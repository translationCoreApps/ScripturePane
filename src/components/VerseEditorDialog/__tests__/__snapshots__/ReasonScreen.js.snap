// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReasonScreen/> has no selections 1`] = `
<div
  style="display:flex"
>
  <div
    style="flex-grow:2"
  />
  <div
    style="display:flex;flex-grow:1"
  >
    <div
      style="flex-grow:1"
    >
      <div>
        <input
          name="spelling"
          type="checkbox"
        />
         editor.spelling
      </div>
      <div>
        <input
          name="punctuation"
          type="checkbox"
        />
         editor.punctuation
      </div>
      <div>
        <input
          name="word_choice"
          type="checkbox"
        />
         editor.word_choice
      </div>
    </div>
    <div
      style="flex-grow:1"
    >
      <div>
        <input
          name="meaning"
          type="checkbox"
        />
         editor.meaning
      </div>
      <div>
        <input
          name="grammar"
          type="checkbox"
        />
         editor.grammar
      </div>
      <div>
        <input
          name="other"
          type="checkbox"
        />
         editor.other
      </div>
    </div>
  </div>
  <div
    style="flex-grow:2"
  />
</div>
`;

exports[`<ReasonScreen/> has selections 1`] = `
<div
  style="display:flex"
>
  <div
    style="flex-grow:2"
  />
  <div
    style="display:flex;flex-grow:1"
  >
    <div
      style="flex-grow:1"
    >
      <div>
        <input
          checked=""
          name="spelling"
          type="checkbox"
        />
         editor.spelling
      </div>
      <div>
        <input
          name="punctuation"
          type="checkbox"
        />
         editor.punctuation
      </div>
      <div>
        <input
          name="word_choice"
          type="checkbox"
        />
         editor.word_choice
      </div>
    </div>
    <div
      style="flex-grow:1"
    >
      <div>
        <input
          name="meaning"
          type="checkbox"
        />
         editor.meaning
      </div>
      <div>
        <input
          name="grammar"
          type="checkbox"
        />
         editor.grammar
      </div>
      <div>
        <input
          name="other"
          type="checkbox"
        />
         editor.other
      </div>
    </div>
  </div>
  <div
    style="flex-grow:2"
  />
</div>
`;
